File: project/jimmer-spring-boot-starter/src/test/java/org/babyfish/jimmer/spring/java/model/BookInput.java
package org.babyfish.jimmer.spring.java.model;

import org.babyfish.jimmer.ImmutableConverter;
import org.babyfish.jimmer.ImmutableObjects;
import org.babyfish.jimmer.client.Doc;
import org.babyfish.jimmer.client.ExportFields;
import org.jetbrains.annotations.Nullable;

import java.math.BigDecimal;
import java.util.List;

@ExportFields
public class BookInput implements Input<Book> {

    private static final ImmutableConverter<Book, BookInput> CONVERTER =
            ImmutableConverter
                    .forFields(Book.class, BookInput.class)
                    .map(
                            BookProps.STORE,
                            "storeId",
                            value -> ImmutableObjects.makeIdOnly(BookStore.class, value)
                    )
                    .mapList(
                            BookProps.AUTHORS,
                            "authorIds",
                            value -> ImmutableObjects.makeIdOnly(Author.class, value)
                    )
                    .build();

    private final String name;

    private final int edition;

    private final BigDecimal price;

    @Doc("Null is allowed")
    @Nullable
    private final Long storeId;

    private final List<Long> authorIds;

    public BookInput(
            String name,
            int edition,
            BigDecimal price,
            @Nullable Long storeId,
            List<Long> authorIds
    ) {
        this.name = name;
        this.edition = edition;
        this.price = price;
        this.storeId = storeId;
        this.authorIds = authorIds;
    }

    @Override
    public Book toEntity() {
        return CONVERTER.convert(this);
    }

    @Override
    public String toString() {
        return "BookInput{" +
                "name='" + name + '\'' +
                ", edition=" + edition +
                ", price=" + price +
                ", storeId=" + storeId +
                ", authorIds=" + authorIds +
                '}';
    }
}

File: example/kotlin/jimmer-sql-kt/src/main/kotlin/org/babyfish/jimmer/example/kt/sql/model/input/BookInput.kt
package org.babyfish.jimmer.example.kt.sql.model.input

import org.babyfish.jimmer.ImmutableConverter
import org.babyfish.jimmer.example.kt.sql.model.Book
import org.babyfish.jimmer.kt.makeIdOnly
import org.babyfish.jimmer.kt.map
import org.babyfish.jimmer.kt.mapList
import org.babyfish.jimmer.spring.java.model.Input
import java.math.BigDecimal

data class BookInput(
    val id: Long?,
    val name: String,
    val edition: Int,
    val price: BigDecimal,
    val storeId: Long?,
    val authorIds: List<Long>
): Input<Book> {

    override fun toEntity(): Book =
        CONVERTER.convert(this)

    companion object {

        @JvmStatic
        private val CONVERTER = ImmutableConverter
            .forFields(Book::class.java, BookInput::class.java)
            .map(Book::id) {
                useIf { it.id !== null }
            }
            .map(Book::store, BookInput::storeId) {
                valueConverter(::makeIdOnly)
            }
            .mapList(Book::authors, BookInput::authorIds) {
                elementConverter(::makeIdOnly)
            }
            .build()
    }
}